src/ProxyProtocol.cc

```diff
@@ -0,0 +1,324 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include <algorithm>
+#include "parser/BinaryTokenizer.h"
+#include "parser/Tokenizer.h"
+#include "ProxyProtocol.h"
+#include "sbuf/StringConvert.h"
+#include "sbuf/Stream.h"
+#include "SquidConfig.h"
+#include "SquidString.h"
+#include "StrList.h"
+
+#if HAVE_SYS_SOCKET_H
+#include <sys/socket.h>
+#endif
+#if HAVE_NETINET_IN_H
+#include <netinet/in.h>
+#endif
+#if HAVE_NETINET_IP_H
+#include <netinet/ip.h>
+#endif
+
+/// magic octet prefix for PROXY protocol version 1
+static const SBuf Proxy1p0magic("PROXY ", 6);
+
+/// magic octet prefix for PROXY protocol version 2
+static const SBuf Proxy2p0magic("\x0D\x0A\x0D\x0A\x00\x0D\x0A\x51\x55\x49\x54\x0A", 12);
+
+ProxyProtocol::Message::Message(const char *ver, const uint16_t len, const uint8_t cmd)
+    : version_(ver), length_(len), command_(Two::CommandType(cmd)) {}
+
+SBuf
+ProxyProtocol::Message::getAll(const char sep) const
+{
+    SBuf result;
+    char ipBuf[MAX_IPSTRLEN];
+    result.appendf(":version: %s\r\n", version_);
+    result.appendf(":command: %d\r\n", command_);
+    auto srcAddr = srcIpAddr;
+    srcAddr.applyClientMask(Config.Addrs.client_netmask);
+    result.appendf(":src_addr: %s\r\n", srcAddr.toStr(ipBuf, MAX_IPSTRLEN));
+    result.appendf(":src_port: %d\r\n", srcIpAddr.port());
+    result.appendf(":dst_addr: %s\r\n", dstIpAddr.toStr(ipBuf, MAX_IPSTRLEN));
+    result.appendf(":dst_port: %d\r\n", dstIpAddr.port());
+    for (const auto &tlv: tlvs) {
+        result.appendf("%d: ", tlv.type);
```

Comment:
Should we use getValues() here instead of (poorly!) duplicating that code? I suspect the answer is "no", but we should then add a comment here explaining why we should not. Here is a rough sketch:
```
// Do not use Message::getValues() because it has to merge same-ID TLV values,
// but we can (and want to) preserve the original PROXY header layout.
```